In [41]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [42]:
data=pd.read_csv("C:\\Users\\attil\\Downloads\\github_illusion_clean.csv")

In [19]:
data.head

<bound method NDFrame.head of         Unnamed: 0          username account_created last_commit_date  \
0                0    wblackburn1825      2021-08-17       2020-10-11   
1                1    reneecampos410      2012-05-25       2022-11-02   
2                2      paullutz4507      2022-07-21       2024-02-03   
3                3     maureen114013      2009-11-02       2023-12-16   
4                4        dayala3658      2015-11-23       2022-11-18   
...            ...               ...             ...              ...   
101802      104995   joebenjamin3839      2016-04-25       2023-08-31   
101803      104996  kennedyangel3873      2012-09-03              NaN   
101804      104997       willie15884      2019-04-20       2023-01-08   
101805      104998       james816175      2013-02-14       2021-01-09   
101806      104999     jessica106446      2022-03-11              NaN   

        total_repos  total_commits  commit_streak  avg_commit_hour  stars  \
0              4

In [20]:
data.dtypes

Unnamed: 0             int64
username              object
account_created       object
last_commit_date      object
total_repos          float64
total_commits        float64
commit_streak        float64
avg_commit_hour      float64
stars                  int64
forks                  int64
star_fork_ratio      float64
followers            float64
following            float64
readme_score         float64
language              object
country               object
has_bio                int64
has_profile_pic        int64
bot_like_score       float64
suspicion_score      float64
profile_age_days     float64
_profile_type          int64
account_age_years    float64
commit_frequency     float64
follow_ratio         float64
repo_commit_ratio    float64
suspicious_hrs         int64
is_suspicious          int64
dtype: object

#### Q1. How many profiles are suspicious vs genuine?

In [21]:
data['is_suspicious'].value_counts()

is_suspicious
0    67247
1    34560
Name: count, dtype: int64

Out of 101807 profiles in the dataset, 67247 were flagged as genuine and 34560 were flagged as suspicious.

#### Q2. What percentage of all profiles are flagged suspicious?

In [22]:
suspicious_pct=data['is_suspicious'].mean()*100
print(f"{suspicious_pct:.2f}% of profiles are suspicious")

33.95% of profiles are suspicious


33-95% of all profiles in the dataset show suspicious, bot-like
activity patterns.

#### Q3. What is the average README score overall, and split by suspicious vs genuine?

In [23]:
data['readme_score'].mean()
data.groupby('is_suspicious')['readme_score'].mean()

is_suspicious
0    6.279919
1    2.277465
Name: readme_score, dtype: float64

On average, genuine profiles score 6.3/10 on README quality, while
suspicious profiles score only 2.3/10 — a gap of 4 points.

#### Q4. What are the top 10 profiles by stars?

In [24]:
data.sort_values('stars',ascending=False)[['username','stars','forks','is_suspicious']].head(10)

,username,stars,forks,is_suspicious
17133,brucemills6998,7559,83,1
92108,davidvillegas6881,7236,8,1
89895,nicole654954,7067,127,1
40178,lmurphy414,6991,109,1
16412,ucolon9283,6917,6,1
75066,anthonyjennifer857,6866,9,1
52152,susancooper6614,6600,14,1
9430,connienguyen2760,6577,42,1
74436,twhitney8626,6314,95,1
3697,nicole277956,6236,118,1


Among the top 10 most-starred profiles, 10 out of 10 are flagged as suspicious, showing that raw star count alone is a poor signal of
quality.

#### Q5. How many profiles have stars > 1000 but forks < 10?

In [25]:
fake_popularity=data[(data['stars']>1000) & (data['forks']<10)]
len(fake_popularity)

2668

2668 profiles (2.6% of the dataset) have over 1,000 stars but fewer than 10 forks — a textbook sign of inflated, unearned popularity.

#### Q6. Of those high-star/low-fork profiles, what % are flagged suspicious?

In [26]:
fake_popularity['is_suspicious'].mean()*100

np.float64(99.36281859070465)

Of these high-star/low-fork profiles, 99.3% are independently
flagged as suspicious by the scoring system -confirming this pattern is a
genuine fake-popularity signal, not a coincidence.

#### Q7. What percentage of profiles commit during "suspicious hours" (suspicious_hrs == 1)?

In [27]:
data.groupby('is_suspicious')['suspicious_hrs'].mean()*100

is_suspicious
0     0.063943
1    66.258681
Name: suspicious_hrs, dtype: float64

Within suspicious profiles this rises to 66.26%, vs only 0.064% for genuine ones — a ~1036× difference.

#### Q8. What is the average follow_ratio for suspicious vs genuine profiles?

In [28]:
data.groupby('is_suspicious')['follow_ratio'].mean()

is_suspicious
0    13.411706
1     2.363233
Name: follow_ratio, dtype: float64

Suspicious profiles have an average follow ratio of 2.36, compared to 13.41 for genuine profiles — meaning suspicious accounts get followed back 5.68 times less, relative to how many accounts they follow.

#### Q9. How many profiles have a follow_ratio below 0.05 (extreme mass-follow behavior)?

In [29]:
mass_followers=data[data['follow_ratio']<0.05]
len(mass_followers)

# % of these are suspicious
mass_followers['is_suspicious'].mean()*100

np.float64(97.60623229461757)

28240 profiles (27.7% of the dataset) follow far more accounts than follow them back. Of these, 97% are already flagged as suspicious.

#### Q10. Which country has the most profiles in the dataset?

In [30]:
data['country'].value_counts().head(10)

country
United States     12583
Unknown           10230
India             10111
Brazil             7785
Japan              7684
Canada             7675
United Kingdom     7660
China              7650
Germany            7625
Australia          7622
Name: count, dtype: int64

United States has the most profiles in the dataset at 12583, followed
by India with 10111.

#### Q11. Which country has the highest average suspicion_score?

In [31]:
data.groupby('country')['suspicion_score'].mean().sort_values(ascending=False).head(10)

country
Australia         0.375536
United Kingdom    0.372210
India             0.371427
Russia            0.369595
United States     0.369458
Canada            0.369431
Brazil            0.369420
China             0.368964
Germany           0.367818
France            0.367453
Name: suspicion_score, dtype: float64

Australia has the highest average suspicion score at 0.378,
notably above the overall dataset average of 0.368.

#### Q12. Which programming language has the highest average stars?

In [32]:
data.groupby('language')['stars'].mean().sort_values(ascending=False).head(10)

language
Typescript    344.257863
Rust          343.630009
R             343.177061
Kotlin        342.798266
Go            338.114938
Unknown       337.131697
Php           336.742934
Shell         333.479371
Javascript    333.063186
C++           332.542741
Name: stars, dtype: float64

TypeScript developers have the highest average star count at
344.3 — 2.6% above the overall average

#### Q13. Which language has the highest percentage of suspicious profiles? (Only consider languages with at least 500 profiles)

In [33]:
temp=data['language'].value_counts()
valid=temp[temp>=500].index
valid
check=data[data['language'].isin(valid)]
data.groupby('language')['is_suspicious'].mean().sort_values(ascending=False)*100

language
Rust          34.620311
Php           34.527290
Shell         34.417976
Unknown       34.252940
Swift         34.193548
Typescript    34.167262
Go            33.979189
R             33.960573
Javascript    33.883999
Kotlin        33.736681
C++           33.599110
Java          33.539169
Python        33.062783
Ruby          32.999629
Name: is_suspicious, dtype: float64

Among languages with 500+ profiles, Rust has the highest
suspicious rate at 34.6%, compared to the overall average of 33.9%.

#### Q14. Cross-tab: Account age bucket vs Suspicious status

In [34]:
data['age_bucket']=pd.cut(
    data['account_age_years'],
    bins=[0,1,2,4,6,8,20],
    labels=['<1yr','1-2yr','2-4yr','4-6yr','6-8yr','8yr+']
)
pd.crosstab(data['age_bucket'],data['is_suspicious'],normalize='index')*100

is_suspicious,0,1
age_bucket,,
2-4yr,66.243194,33.756806
4-6yr,65.861848,34.138152
6-8yr,65.374134,34.625866
8yr+,66.192251,33.807749


Suspicious activity is fairly consistent across account age groups, ranging from 33.8% to 34.6%, suggesting that account age alone is not a strong predictor of suspicious behavior.

#### Q15. For each suspicion tier (Clean / Borderline / Likely Bot / Confirmed Bot), what's the average README score and average stars?

In [35]:
data['suspicion_tier']=pd.cut(
    data['suspicion_score'],
    bins=[0,0.25,0.50,0.75,1.01],
    labels=['Clean','Borderline','Likely Bot','Confirmed Bot']
)

data.groupby('suspicion_tier')[['readme_score','stars','forks','follow_ratio']].mean()

C:\Users\attil\AppData\Local\Temp\ipykernel_26148\3736933357.py:7: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  data.groupby('suspicion_tier')[['readme_score','stars','forks','follow_ratio']].mean()


,readme_score,stars,forks,follow_ratio
suspicion_tier,,,,
Clean,6.592613,134.856630,38.459897,13.837166
Borderline,5.557658,191.376736,34.071192,12.735928
Likely Bot,2.740620,552.161545,14.927926,4.913065
Confirmed Bot,1.834448,825.013381,7.837063,0.006709


'Confirmed Bot' tier profiles average 825.01 stars and a README score of 1.83/10, compared to 134.86 stars and 6.59/10 for 'Clean' tier profiles — almost a complete reversal in profile quality indicators.

#### Q16. Which 5 profiles look the MOST suspicious overall, and what do they have in common?

In [36]:
top_suspicious=data.sort_values('suspicion_score',ascending=False).head(5)
top_suspicious[['username','stars','forks','readme_score','follow_ratio','commit_streak','suspicious_hrs','suspicion_score']]

,username,stars,forks,readme_score,follow_ratio,commit_streak,suspicious_hrs,suspicion_score
65279,christineburnett8158,273,3,0.8,0.003185,300.0,1,1.0
101714,thomasluke2318,92,1,0.6,0.004981,100.0,1,1.0
65264,gmorrison6356,411,3,1.2,0.002635,365.0,1,1.0
19026,mnichols8252,433,1,0.5,0.004170,500.0,1,1.0
28451,ryoung6694,167,2,2.1,0.003518,200.0,1,1.0


The 5 most suspicious profiles in the dataset share clear traits: an average of 275.2 stars but only 2.0 forks, README scores averaging just 1.04/10, and 100% of them commit during suspicious hours.

#### Q17. For each language, show profile count, average suspicion score, and average README score — all in ONE table.

In [37]:
language_summary=data.groupby('language').agg(
    profile_count=('username','count'),
    avg_suspicion=('suspicion_score','mean'),
    avg_readme=('readme_score','mean')
).sort_values('avg_suspicion',ascending=False)

language_summary

,profile_count,avg_suspicion,avg_readme
language,,,
Shell,8144,0.372772,4.897151
Swift,5425,0.372141,4.903705
Unknown,13774,0.372071,4.928372
Rust,5465,0.371101,4.874657
Php,8208,0.371082,4.926693
Kotlin,5537,0.369087,4.906357
Javascript,8293,0.369010,4.941613
Typescript,5596,0.368722,4.908113
C++,5393,0.367368,4.933914


Shell profiles have the highest average suspicion score (0.373) and one of the lowest average README scores (4.90/10), indicating a higher concentration of suspicious characteristics compared to other language communities.

#### Q18. Build a table: for the top 10 countries, how many profiles fall into each suspicion tier (Clean/Borderline/Likely Bot/Confirmed Bot)?

In [40]:
top_countries=data['country'].value_counts().head(10).index
filtered=data[data['country'].isin(top_countries)]

table=pd.pivot_table(
    filtered,
    index='country',
    columns='suspicion_tier',
    values='username',
    aggfunc='count',
    fill_value=0
)

table

C:\Users\attil\AppData\Local\Temp\ipykernel_26148\3917428215.py:4: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  table=pd.pivot_table(


suspicion_tier,Clean,Borderline,Likely Bot,Confirmed Bot
country,,,,
Australia,2962,1467,1287,1361
Brazil,3165,1430,1228,1405
Canada,3143,1425,1237,1361
China,3110,1479,1243,1325
Germany,3097,1418,1186,1367
India,4075,1949,1617,1804
Japan,3095,1470,1276,1261
United Kingdom,3073,1476,1229,1372
United States,5134,2394,2060,2184


The distribution of 'Confirmed Bot' profiles is relatively consistent across major countries, ranging from 17.76% to 19.44%. Brazil records the highest share, while Japan has the lowest, indicating only modest geographic variation in suspicious activity.

#### Q19. Split profiles into Low/Medium/High commit frequency groups, then check the suspicious % in each group.

In [80]:
low_cutoff=data['commit_frequency'].quantile(0.33)
high_cutoff=data['commit_frequency'].quantile(0.66)

def freq_bucket(x):
    if(x<low_cutoff):
        return 'Low'
    elif(x<high_cutoff):
        return 'Medium'
    else:
        return 'High'

data['freq_level']=data['commit_frequency'].apply(freq_bucket)

data.groupby('freq_level')['is_suspicious'].mean()*100

freq_level
High      71.015456
Low       11.944874
Medium    17.755090
Name: is_suspicious, dtype: float64

High-frequency committers are nearly 6× more likely to be classified as suspicious than low-frequency committers.

#### Q20. How many profiles trigger ALL FOUR major red flags at once — and does your suspicion_score system already catch them?

In [82]:
red_flags=(
    (data['stars']>500) &
    (data['follow_ratio']<0.05) &
    (data['readme_score']<3) &
    (data['suspicious_hrs']==1)
)

extreme_profiles=data[red_flags]
print(f"Profiles mathcing all 4 red flags: {len(extreme_profiles)}")
print(f"% already flagged as suspicious: {extreme_profiles['is_suspicious'].mean()*100:.1f}%")

Profiles mathcing all 4 red flags: 9124
% already flagged as suspicious: 100.0%


9,124 profiles trigger all four major red flags simultaneously, and 100.0% of them were already caught by the suspicion score system — validating that the scoring logic successfully identifies the most extreme suspicious profiles.